### PR TITLE
Use hdf5 file to load catalog data

### PR DESCRIPTION
**Pull request by ahuang314**
Replaces np.load with h5py.File to open an hdf5 file
Added a dummy weighted catalog file to the repo so that testing can be done. This dummy weighted catalog file contains data for 5 galaxies.

### CODE REVIEW

File: paltax/InputConfigs/input_config.py
Comment (swagnercarena) [inline paltax/InputConfigs/input_config.py:108]: Please keep this in the cosmos folder

File: paltax/InputConfigs/input_config.py
Comment (swagnercarena) [inline paltax/InputConfigs/input_config.py:114]: I think the [:] can be replaced with a .value. Also the formatting is a bit odd here. Why are the images saved in galaxy_images directly whereas the parameters and weights first need to be accessed through 'parameters_and_weights'. I think it would be simpler to:

1. Have all the images, parameters, and weights be keys of the hdf5_file (i.e. directly access gini_weights or gini)
2. Unless there's a very specific reason, I would prefer that all the keys be lowercase. For example, `gini` and `gini_weights` instead of `GINI` as you have it now. Listing all of the options in the CONFIG is a nice touch though.

File: paltax/InputConfigs/input_config.py
Comment (swagnercarena) [inline paltax/InputConfigs/input_config.py:121]: We don't want to be using both source models, just the new weightedcatalog model

File: paltax/InputConfigs/input_config_br.py
Comment (swagnercarena) [inline paltax/InputConfigs/input_config_br.py:120]: See comments on input_config file

File: paltax/InputConfigs/input_config_br.py
Comment (swagnercarena) [inline paltax/InputConfigs/input_config_br.py:112]: See comments on input_config file

File: paltax/InputConfigs/input_config_psf.py
Comment (swagnercarena) [inline paltax/InputConfigs/input_config_psf.py:112]: See comments on input_config file

File: paltax/InputConfigs/input_config_psf.py
Comment (swagnercarena) [inline paltax/InputConfigs/input_config_psf.py:121]: See comments on input_config file

File: paltax/InputConfigs/input_config_test.py
Comment (swagnercarena) [inline paltax/InputConfigs/input_config_test.py:113]: See comments on input_config file

File: paltax/InputConfigs/input_config_test.py
Comment (swagnercarena) [inline paltax/InputConfigs/input_config_test.py:121]: See comments on input_config file

File: paltax/source_models.py
Comment (swagnercarena) [inline paltax/source_models.py:26]: put in alphabetic order

File: paltax/source_models.py
Comment (swagnercarena) [inline paltax/source_models.py:299]: The first [:] we'll probably want to modify to access the current batch of images. After that, I think we want to use .value (avoid the other two [:])

Comment (swagnercarena) [reply paltax/source_models.py:299]: I leave this up to you, but I think just using the label `images` might be more standard than specifying `galaxy_images`.

File: paltax/source_models.py
Comment (swagnercarena) [inline paltax/source_models.py:301]: .value here and below

File: paltax/InputConfigs/input_config.py
Comment (swagnercarena) [inline paltax/InputConfigs/input_config.py:110]: Here an in other init files, use with to open the hdf5 file and automatically close it once the weights are extracted

File: paltax/source_models.py
Comment (swagnercarena) [inline paltax/source_models.py:298]: Since we're switching to hdf5, I propose three changes.

1. The cosmos catalog should now include a parameter that determines how many images to load when modify_cosmology_params is called. It should have a relatively large default value (say 2048).
2. The hdf5_file should be opened in init and saved as a self property. You should use [atexit](https://docs.python.org/3/library/atexit.html#atexit.register) at the end of the __init__ function to register a cleanup function that will close the hdf5 file once python exits. 
3. You should save a tracker of the number of times modify_cosmology_params has been called and the total length of the hdf5 file. The idea here is that each time we call modify_cosmology_params we should be accessing a new chunk of the hdf5 file. When we run out of chunks, we should start again at the first chunk. Happy to talk on slack if you need more details on this.

The high-level idea here is to make sure that we can use hdf5 files that are larger than the number of images we can realistically store on memory. Note that once you've made that change, we'll also have to change how you currently deal with the weights.

**Review by swagnercarena — CHANGES_REQUESTED** [review]

File: paltax/InputConfigs/input_config.py
Comment (ahuang314) [reply paltax/InputConfigs/input_config.py:114]: It made the most sense to me to group all of the parameters and weights into a common "parameters_and_weights" group, rather than having 16 keys right from the get-go, one for each parameter/weight. But yes, I can change it, these should be easy fixes

File: paltax/source_models.py
Comment (ahuang314) [reply paltax/source_models.py:301]: Do you mean like this? 
image
It doesn't work:
image
I also tried .value(), .values, and .values()

Comment (swagnercarena) [reply paltax/source_models.py:301]: Looks like the .value was deprecated in version 2.1, so you can ignore those comments. Anyways once we implement the ability to load a portion of the dataset the [:] notation would be the appropriate choice.

**Comment by ahuang314** [comment]
@swagnercarena OK, I've implemented most of your suggested changes plus some more. Here's what I've done:

1. `WeightedCatalog` no longer takes `catalog_weights` as input during initialization. Now it takes just a paramater (e.g. 'gini'), and the catalog weights are loaded up during the `__init__` function instead. This way we do not have to open the hdf5 file in the input config.
2. Added functionality to load batches of the images at a time. I added a `batch_size` parameter to the `CosmosCatalog` as you suggested. `self.batch_number` keeps track of the number of times `modify_cosmology_params` has been called. `self.total_num_galaxies` keeps track of the size of the hdf5 file so we know when to loop back to the first batch.
3. hdf5 formatting changed as you suggested
4. I have not yet changed any of the test files to accommodate the new structure of the classes. Will work on that once you approve of the new structure

File: paltax/source_models.py
Comment (swagnercarena) [reply paltax/source_models.py:316]: Also I think you changed the name to images

File: paltax/source_models.py
Comment (swagnercarena) [inline paltax/source_models.py:466]: Just save the parameter. Load and process the weights later.

**Review by swagnercarena — CHANGES_REQUESTED** [review]

File: paltax/InputConfigs/input_config.py
Comment (ahuang314) [reply paltax/InputConfigs/input_config.py:107]: Ah, gotcha. Yeah, I had gotten confused from the previous comment

File: paltax/InputConfigs/input_config.py
Comment (ahuang314) [reply paltax/InputConfigs/input_config.py:111]: Whoops. What happened was I made changes to one of the config files and copy pasted the snippet of code to the other config files, not realizing that they were different.

File: notebooks/GenerateImages.ipynb
Comment (ahuang314) [reply notebooks/GenerateImages.ipynb:1]: Yes, I changed this notebook intentionally because it had that cell that was not supposed to be there as you mentioned

File: paltax/source_models.py
Comment (swagnercarena) [inline paltax/source_models.py:469]: You should save the parameter value, not the catalog_weights. Technically this would also work, but I think it would be clearer if we accessed the hdf5 file in the modify_cosmology_params call.

File: paltax/source_models.py
Comment (swagnercarena) [inline paltax/source_models.py:472]: These needs to be done on the chunks drawn in modify_cosmology_params

**Review by swagnercarena — CHANGES_REQUESTED** [review]

**Review by swagnercarena — CHANGES_REQUESTED** [review]
Looks good, now we just need to modify the tests.

1. Make sure all the old tests are passing. If the failure of a test is confusing to you, we can discuss it on slack / on this thread.
2. We should write a new test to make sure the chunking is well behaved. Essentially we should call modify_cosmology_params enough to loop through the test dataset and confirm the the images we expect to be loaded are the ones that are loaded in each call (i.e. image 0:2, 2:4, 4:5, 0:2, ... if the chunk size is 2 and the dataset has 5 images).

File: paltax/source_models_test.py
Comment (swagnercarena) [inline paltax/source_models_test.py:37]: I like the inclusion of this note, but put it closer to where the tests make this assumption instead of the top of the file.

File: paltax/source_models_test.py
Comment (swagnercarena) [inline paltax/source_models_test.py:58]: put _ in front of these functions to clarify that they are only for internal use within this test file.

File: paltax/source_models_test.py
Comment (swagnercarena) [inline paltax/source_models_test.py:64]: See above comment.

File: paltax/source_models_test.py
Comment (swagnercarena) [inline paltax/source_models_test.py:317]: For consistency, can you use self.assertEqual here and in other places where you use assert

File: paltax/source_models_test.py
Comment (swagnercarena) [inline paltax/source_models_test.py:319]: CosmosCatalog not WeightedCatalog

File: paltax/source_models_test.py
Comment (swagnercarena) [inline paltax/source_models_test.py:338]: see comment on assert

File: paltax/source_models_test.py
Comment (swagnercarena) [inline paltax/source_models_test.py:492]: see comment on assert

File: paltax/source_models_test.py
Comment (swagnercarena) [inline paltax/source_models_test.py:518]: see comment on assert

File: paltax/source_models_test.py
Comment (swagnercarena) [inline paltax/source_models_test.py:544]: see comment on assert

**Review by swagnercarena — CHANGES_REQUESTED** [review]
Looks really good. Asked for some small changes but should be good to go after that.

**Comment by swagnercarena** [comment]
It looks like the coverage on source_models.py [has dropped](https://coveralls.io/builds/69083919). Would you mind running coverage locally to see what line is missing? 

```
coverage run --source paltax --omit="*Configs/*,*_test.py,main.py" -m unittest discover -v -p "*_test.py"
coverage html
open html/index.rst
```

**Review by swagnercarena — APPROVED** [review]

**Pull request merged**
